plutus-metatheory/src/Signature.lagda.md

```diff
@@ -0,0 +1,170 @@
+
+```
+record Sig : Set where 
+   constructor sig
+   field
+     -- number of type variables
+    fv♯ : ℕ   
+     -- list of arguments
+    args : Args fv♯
+     -- type of result
+    result : fv♯ ⊢♯
+
+```
+
+## Obtaining concrete types from signatures
+
+Signatures represent abstract types which need to be made concrete in 
+order to use them. The following modules may be instantiated to obtain 
+a function `sig2type` which converts from an abstract into a concrete type.
+
+The instantation is defined in two stages. In the first stage, the following parameters are instantiated:
+
+   1. the set `Con` over which types are indexed,
+   2. the types `Ty` (indexed over `Con`),
+   3. a function interpreting naturals as elements of `Con`, and 
+   4. a function interpreting variable indexes into types in `Ty.
+
+```
+module FromSig1 (Con : Set)(Ty : Con → Set) 
+               (nat2Con : ℕ → Con) 
+               (fin2Ty : ∀{n} → Fin n → Ty (nat2Con n))
+               where
+
+```
+
+With these parameters instantiated we can bring into scope type constants.
+  
+```
+  import Builtin.Constant.Type Con Ty as T2
+  open T2.TyCon
+```
+
+The second stage is instantiating the following parameters:
+  1. a way to insert a type constant as a type,
+  2. the constructor for function types, and
+  3. the constructor for Π types.
+```
+  module FromSig2 
+           (mkTyCon : ∀{n} → T2.TyCon (nat2Con n) → Ty (nat2Con n)) 
+           (_⇒_ : ∀{n} → Ty (nat2Con n) → Ty (nat2Con n) → Ty (nat2Con n))
+           (Π : ∀{n} → Ty (nat2Con (suc n)) → Ty (nat2Con n))
+          where
+
+    -- convert a built-in-compatible type into a Ty type
+    ♯2* : ∀{n} → n ⊢♯ → Ty (nat2Con n)
+    ♯2* (` x) = fin2Ty x
+    ♯2* (con integer) = mkTyCon T2.integer
+    ♯2* (con bytestring) = mkTyCon  T2.bytestring
+    ♯2* (con string) = mkTyCon  T2.string
+    ♯2* (con unit) = mkTyCon  T2.unit
+    ♯2* (con bool) = mkTyCon  T2.bool
+    ♯2* (con (list x)) = mkTyCon  (T2.list (♯2* x))
+    ♯2* (con (pair x y)) = mkTyCon  (T2.pair (♯2* x) (♯2* y))
+    ♯2* (con Data) = mkTyCon  T2.Data
+
+    {- `sig2type-aux` takes a list of arguments and a result type, and produces
+        a function that takes all arguments and returns the result type.
+
+       sig2type-aux [ b₁ , b₂ , ... ,bₙ ]  tᵣ = t₁ ⇒ t₂ ⇒ ... ⇒ tₙ ⇒ tᵣ
+          where tᵢ = ♯2* bᵢ
+    -}
+    sig2type-aux : ∀{n} 
+              → Args n 
+              → Ty (nat2Con n) → Ty (nat2Con n)
+    sig2type-aux xs = foldr (λ A xs res → (♯2* A) ⇒ (xs res)) (λ A res → ♯2* A ⇒ res) xs
+
+    -- take a type and close its type variables using Π
+    sig2type-aux2 : ∀{n} → Ty (nat2Con n) → Ty (nat2Con 0)
+    sig2type-aux2 {zero} t = t
+    sig2type-aux2 {suc n} t = sig2type-aux2 {n} (Π t)
+
+    -- The main conversion function
+    sig2type : Sig → Ty (nat2Con 0)
+    sig2type (sig _ as res) = sig2type-aux2 (sig2type-aux as (♯2* res)) 
+```
+
+The parameter `Con` above is usually `Ctx⋆`.  In this case, the parameters
```

Comment:
This may be confusing as we also have 'con'structors

Reply:
I was taking it from the `Builtin.Constant.Type` module. I've changed to `Ctx` (only in this module).